### PR TITLE
Add query pruning middleware

### PR DESCRIPTION
**Pull request by zenador**
#### What this PR does

(Hackathon project) Query pruning optimises PromQL queries by rewriting them to avoid evaluating expressions that can't produce a result, which is particularly useful in [dashboards with toggles between native and classic histograms](https://github.com/grafana/jsonnet-libs/pull/1164/files#diff-040bce91f6d46ff86f067fe7e11934183b2ea095ff6d4c62c5e798f24b6d9269R86).

#### Which issue(s) this PR fixes or relates to

N/A

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by tacole02 — APPROVED** [review]
LGTM 😄

**Review by krajorama — COMMENTED** [review]
awesome work, I only have  a couple of minor comments

File: pkg/frontend/querymiddleware/prune_test.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	seriesName := `test_float`
+	for i := 0; i < numSeries; i++ {
+		nss := promql.NewStorageSeries(promql.Series{
+			Metric:     labels.FromStrings("__name__", seriesName, "series", fmt.Sprint(i)),
+			Floats:     floats,
+			Histograms: histograms,
+		})
+		storageSeries = append(storageSeries, nss)
+	}
+	queryable := storageSeriesQueryable(storageSeries)
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/prune_test.go:42]: `promqltest.LoadedStorage` might help simplify this setup

File: pkg/frontend/querymiddleware/roundtrip.go [hunk 2]
```diff
@@ -65,6 +65,7 @@ type Config struct {
 	MaxRetries               int           `yaml:"max_retries" category:"advanced"`
 	NotRunningTimeout        time.Duration `yaml:"not_running_timeout" category:"advanced"`
 	ShardedQueries           bool          `yaml:"parallelize_shardable_queries"`
+	PrunedQueries            bool          `yaml:"prune_queries"`
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/roundtrip.go:68]: Should we mark this as experimental?

File: pkg/frontend/querymiddleware/roundtrip.go [hunk 3]
```diff
@@ -92,6 +93,7 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.DurationVar(&cfg.SplitQueriesByInterval, "query-frontend.split-queries-by-interval", 24*time.Hour, "Split range queries by an interval and execute in parallel. You should use a multiple of 24 hours to optimize querying blocks. 0 to disable it.")
 	f.BoolVar(&cfg.CacheResults, "query-frontend.cache-results", false, "Cache query results.")
 	f.BoolVar(&cfg.ShardedQueries, "query-frontend.parallelize-shardable-queries", false, "True to enable query sharding.")
+	f.BoolVar(&cfg.PrunedQueries, "query-frontend.prune-queries", false, "True to enable query pruning.")
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/roundtrip.go:96]: [nit] I'm not sure "pruning" makes this feature clear enough - what about something like this?

```suggestion
	f.BoolVar(&cfg.PrunedQueries, "query-frontend.prune-queries", false, "True to enable pruning dead code (eg. expressions that cannot produce any results) and simplifying expressions (eg. expressions that can be evaluated immediately) in queries.")
```

File: pkg/frontend/querymiddleware/prune.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	log := spanlogger.FromContext(ctx, p.logger)
+
+	prunedQuery, err := p.pruneQuery(ctx, r.GetQuery())
+	if err != nil {
+		level.Warn(log).Log("msg", "failed to prune the input query, falling back to the original query", "query", r.GetQuery(), "err", err)
+
+		return p.next.Do(ctx, r)
+	}
+
+	level.Debug(log).Log("msg", "query has been purged", "original", r.GetQuery(), "rewritten", prunedQuery)
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/prune.go:46]:
```suggestion
	level.Debug(log).Log("msg", "query has been rewritten by pruning", "original", r.GetQuery(), "rewritten", prunedQuery)
```

Comment (charleskorn) [inline pkg/frontend/querymiddleware/prune.go:46]: [nit] It'd be nice to skip this log message if the expression is not modified (or emit a different message if the expression is not modified)

File: pkg/frontend/querymiddleware/astmapper/pruning.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	case parser.GTR, parser.LSS:
+		return pruner.handleCompOp(expr), false, nil
+	case parser.LOR:
+		return pruner.handleOrOp(expr), false, nil
+	default:
+		return expr, false, nil
+	}
+}
+
+func calcInf(sign bool, coeff string) (*parser.NumberLiteral, bool) {
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/astmapper/pruning.go:59]: [nit] `isPositive` might be a clearer name than `sign`

File: pkg/frontend/querymiddleware/astmapper/pruning.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+	}
+}
+
+func (pruner *queryPruner) pruneBinOp(expr *parser.BinaryExpr) (mapped parser.Expr, finished bool, err error) {
+	switch expr.Op {
+	case parser.MUL:
+		return pruner.handleMultiplyOp(expr), false, nil
+	case parser.GTR, parser.LSS:
+		return pruner.handleCompOp(expr), false, nil
+	case parser.LOR:
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/astmapper/pruning.go:52]: Might be something for another PR: we can do something similar for `and` and `unless`:

* if either side of an `and` is empty, then the whole op will return no results
* if the left side of an `unless` is empty, then the whole op will return no results
* if the right side of an `unless` is empty, then we can just use the left side and ignore the right side

**Review by charleskorn — COMMENTED** [review]
Thanks for putting this together!

**Comment by krajorama** [comment]
Solves https://github.com/grafana/mimir/issues/9065

**Review by charleskorn — APPROVED** [review]
LGTM 🚀

File: pkg/frontend/querymiddleware/prune_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		{`(((-1 * -Inf) < avg(rate(foo[3m]))) unless avg(rate(foo[2m]))) or avg(rate(%s[1m]))`, true},
+		{`((((-1 * -Inf) < avg(rate(foo[4m]))) unless avg(rate(foo[3m]))) and avg(rate(foo[2m]))) or avg(rate(%s[1m]))`, true},
+	}
+	for _, template := range templates {
+		t.Run(template.query, func(t *testing.T) {
+			query := fmt.Sprintf(template.query, seriesName)
+			req := &PrometheusRangeQueryRequest{
+				path:      "/query_range",
+				start:     0,
+				end:       int64(endTime * 1000 * 60),
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/prune_test.go:65]: [nit] This might be clearer as something like `(100 * time.Minute).Milliseconds()`
